File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int low = 0, high = 0, size = s.size(), windowSize, result = 0;
        unordered_map<char,int> letters;
        while(high < size){
            if(letters[s[high]] > 0) letters[s[high]]++;
            else letters[s[high]] = 1;
            windowSize = high - low + 1;
            if(letters.size() == windowSize) result = max(result, windowSize);
            else{
                while(letters.size() < windowSize){
                    if(--letters[s[low]] == 0) letters.erase(s[low]);
                    low++;
                    windowSize = high - low + 1;
                    }
                }
            high++;
    }
        return result;
        }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        charPos = {}
        l, r, maxLength = 0, 0, 0

        while r < len(s):
            if s[r] not in charPos:
                charPos[s[r]] = r
                maxLength = max(maxLength, r - l + 1)
                r += 1
            else: 
                getPos = charPos[s[r]]
                while l <= getPos:
                    del charPos[s[l]]
                    l += 1

        return maxLength